docs/StaticTypingGuide.md

```diff
@@ -0,0 +1,193 @@
+
+```
+target :default do
+  signature "sig"           # ALL signatures from this directory will be loaded
+
+  check "lib/foo/bar"       # ONLY this source code folder will be checked against, using ALL signatures above
+  ignore "lib/foo/bar/baz"  # EXCEPT this subfolder
+end
+```
+
+### Dependency signatures
+
+Steep starts with a [minimal core](https://github.com/ruby/rbs/tree/master/core) loaded type signatures. Adding more [types from the Ruby stdlib](https://github.com/ruby/rbs/tree/master/stdlib) should be done progressively as required:
+
+```
+  library "set" # adds typing for Ruby stdlib's Set
+```
+
+Note: These signatures are part of [`rbs`](), which is included in Ruby releases since Ruby 3.0.
+
+Gems can embed a `sig` directory, which can be used directly:
+
+```
+  library "some_gem_with_a_sig_dir"
+```
+
+Some gems don't have typing information.
+
+In addition, a [vast collection of gems](https://github.com/ruby/gem_rbs_collection) have been typed. These can be fetched via a Rubygems/Bundler-like feature of RBS called [collections](https://github.com/ruby/rbs/blob/e91be7275f4005b1aeac8eadc2faa2b4ad5fdfef/docs/collection.md)
+
+```
+  collection_config "rbs_collection.steep.yaml"
+```
+
+This yaml file is akin to a Gemfile, describes the sources and gem signatures to fetch, and also has a lockfile mechanism. It can also integrate with `bundler` to match the signatures with the gem versions in use.
+
+Otherwise signatures can be vendored:
+
+```
+  repo_path "vendor/rbs"
+  library "subdir"
+```
+
+Typically these are be written as needed for gems entirely missing signatures, and ideally contributed back either upstream to the gem project itself or to the gem rbs collection project.
+
+### Measuring progress
+
+With the described layout and 1:1 match, it becomes easy to track coarse-grained coverage, additions, removals, changes through refactorings, in a similar way as is usually done with unit tests or specs.
+
+In addition, to output typing detailed coverage statistics:
+
+```
+bundle exec steep stats
+```
+
+## Typing a file
+
+### Basics
+
+To type a `.rb` file without a matching `.rbs` file, start with the skeleton:
+
+```
+mkdir -p sig/foo
+bundle exec rbs prototype rb lib/foo/bar.rb > sig/foo/bar.rbs
+```
+
+One can then proceed to [adjusting the signatures](https://github.com/ruby/rbs/blob/e91be7275f4005b1aeac8eadc2faa2b4ad5fdfef/docs/syntax.md) ([by example](https://github.com/ruby/rbs/blob/e91be7275f4005b1aeac8eadc2faa2b4ad5fdfef/docs/rbs_by_example.md)), removing as much `untyped` as possible.
+
+### Type profiling
+
+To discover types, one can leverage [`typeprof`](https://github.com/ruby/typeprof). Contrary to `rbs prototype rb` which relies solely on static parsing, `typeprof` is a Ruby interpreter, except it doesn't *execute* Ruby code, merely evaluates it to track types. Entry point calls to explore the various codepaths are required.
+
+With this file:
```

Comment:
Do you know if there is any tool that gets types based on actual runs, not just abstract runs?

I'm thinking it'd be cool to have something like that to help me type the profiler native stuff. Otherwise, I'll just do it manually -- it's a bit of work but I don't have particularly complex types going in/out of native anyway.

Reply:
There's `rbs prototype runtime` but turns out I never actually used it so I'm not entirely sure.

It's important to note that these are helpful in understanding typing and bootstrapping types, but ultimately it has to be reviewed by a human.